[Storage] WIP / proposal: provide hook for configuring dom fallback for webworkers

maorleger opened the pull request:
This is _super_ early draft - since I'm not confident about the approach here and have little context in this area I wanted to get a draft open to discuss if this is even the approach we want to take. 

We can have the discussion here, and I'll take that into another PR that will include tests, remove the repro project, and clean up the code.

Questions:

1. Is the overall approach reasonable or do we want to go in a different direction?
2. What's a better way to declare the types? I just kind of played with it until things worked 
3. Should the lazy be implemented in both core-http and core-xml, or just core-xml - how far is storage-blob from going to v2?
4. Should the provideFallback be implemented in storage-blob or in core?


High level:

1. Make instantiating DOM objects lazy to allow importing storage-blob without errors
2. Provide a hook to polyfill with whatever DOM implementation you need for web workers

maorleger commented on sdk/storage/storage-blob/src/index.browser.ts:
```diff
@@ -37,3 +37,17 @@ export {
+  document: Document,
+  node: { new (): Node }
+) {
+  if (!self) {
```
What other changes should I do?

- If `window` is defined maybe we don't allow fallbacks?
- spread operator didn't quite work for me here because my Node type is missing the required properties... any ideas what I can do there?
The error I got was:

```
Type '{ new (): Node; TEXT_NODE: number; prototype: Node; }' is missing the following properties from type '{ new (): Node; prototype: Node; readonly ATTRIBUTE_NODE: number; readonly CDATA_SECTION_NODE: number; readonly COMMENT_NODE: number; readonly DOCUMENT_FRAGMENT_NODE: number; ... 13 more ...; readonly TEXT_NODE: number; }': ATTRIBUTE_NODE, CDATA_SECTION_NODE, COMMENT_NODE, DOCUMENT_FRAGMENT_NODE, and 13 more.ts(2740)
```

I'm sure I'm just declaring the types incorrectly, but rather than wrestle with it I figured I'll ask if someone here knows how to declare the type correctly...

maorleger commented:
I pinged a few folks who've been involved in the original discussions around this but feel free to ping anyone you think will be interested?

bterlson commented on sdk/core/core-http/src/util/xml.browser.ts:
```diff
@@ -3,18 +3,43 @@
+ * Simple Lazy that allows for deferred value creation.
+ * In the event of race conditions the last write wins.
+ */
+export class Lazy<T> {
```
This class seems maybe overkill, can you wrap anything needing to be lazy in a function instead? E.g. have `getParser()`, `getNode()`, etc. functions?

maorleger replied:
😄 got a little too excited / ahead of myself

bterlson commented:
I had a chat with @xirzec, and incidentally this is a sort of similar situation that @deyaaeldeen was investigating.

I wonder if this is a case where we can ask users to install a polyfill for the necessary DOM utilities before they import our storage library (or, we keep the lazy, and require the polyfill be installed at some point before the first time we encounter xml)? I worry about having two xml parsing libraries in the dependency when one of them will be used rarely by comparison to browsers and node.

At the limit, we probably don't want to build in support for every runtime platform. To take a stab at some principles/guidelines:

* Provide an out-of-box experience for node and browsers.
* Additional runtimes can be supported out-of-box when
  - code doesn't need to be heavily tailored
  - it can be accomplished using dynamic environment checks/file mappers (e.g. browser/react-native overrides)
  - missing functionality can be provided by relatively tiny shims/polyfills.
* We can light up an out of box experience for a different runtime (and pay the dependency cost) if we find it becomes a common scenario, or also consider providing a different package.

How does this strike your fancy? @maorleger could you maybe see what that sample would look like if we asked users to install the polyfills themselves?

maorleger commented:
@bterlson that all makes sense to me.

To clarify my understanding:
I think my sample changes very little (unless I completely misunderstood) - instead of exposing a method called `configureFallback` I (as the library's consumer / user) inline some similar code before calling the library yea? (I'll also push an update that shows that).

Assuming the above is correct...

The only thing I like about having an explicit fallback method that we own is that we can declare in the signature: here are the dependencies you must polyfill before you can use this library. If we have consumers install them to the global env it becomes an implicit dependency that they might get wrong or miss something. Which is probably fine, but just wanted to bring that up.

Either way I think it's better to start lazy loading these things because imports / requires (at least in my experiments) are hoisted so without the lazy this will unfortunately not work unless you're installing / injecting a polyfill as an import.

```
configureFallback(...)
import storageBlob, { configureFallback } from "@azure/storage-blob";
```

bterlson commented:
My understanding is that both requires and imports are executed in order so you should always be able ensure a polyfill runs before our library code.

One argument in favor of the non-lazy approach is that you get an early error when you didn't polyfill correctly, rather than an error only after encountering xml. What do you think?

jeremymeng commented:
> One argument in favor of the non-lazy approach is that you get an early error when you didn't polyfill correctly, rather than an error only after encountering xml. What do you think?

would it be helpful to throw an error suggesting polyfill when we detect those API are undefined?

maorleger commented:
> My understanding is that both requires and imports are executed in order so you should always be able ensure a polyfill runs before our library code.
> 
> One argument in favor of the non-lazy approach is that you get an early error when you didn't polyfill correctly, rather than an error only after encountering xml. What do you think?

Yeah, that's a good point and probably will lead to more reliable errors in the general case - so I'm on board! 



> > One argument in favor of the non-lazy approach is that you get an early error when you didn't polyfill correctly, rather than an error only after encountering xml. What do you think?
> 
> would it be helpful to throw an error suggesting polyfill when we detect those API are undefined?

I always liked when the error tells me a little more than "document is undefined" so I would like this as well. Anything that specifies: we rely on _Foo_ to be available globally, but it's currently undefined. In certain environments you might need a polyfill before using this library. (Hopefully not as verbose though 😄)


I updated my sample with my own polyfill and reverted all library code to `master` and it works! does this look a little more reasonable? If so, we can do the following:

1. pre-validate and throw a meaningful error for all xml dependencies
2. add a sample of polyfilling using JSDom or something? I copied this sample verbatim from the issue so I would probably rewrite it.

Let me know!
